### PR TITLE
kernel: cleaner solution to pipes SMP race condition

### PR DESCRIPTION
**Pull request by peter-mitsis**
An earlier set of commits had a rather hacky solution to the following race condition in the kernel pipes.

> While processing the list of threads involved in the pipe operation, one or more of those threads timeout could expire, which in turn would lead to the thread waking and executing while data was still being copied to it.

This set of commits proposes a (hopefully) cleaner solution. Waiting threads involved in the data copying are tagged via the k_thread `no_wake_on_timeout` field. This prevents them from timing out while the operation is in progress. This approach also requires some minor post-work cleanup to undo `no_wake_on_timeout` changes in the case of a failure and incomplete timeouts flagged by `timeout_expired`.

### CODE REVIEW

**Comment by MaureenHelm** [comment]
@andyross can you take a look?

File: kernel/pipes.c [hunk 1]
```diff
@@ -220,13 +220,58 @@ static int pipe_walk_op(struct k_thread *thread, void *data)
...
+ * arise from the use of the [no_wake_on_timeout] and [timeout_expired] fields.
+ *
+ * 1. In the event that the request could not be satisfied, the wait queue is
+ *    re-walked to undo the setting of the [no_wake_on_timeout] field. Note
+ *    if a thread's timeout had expired in this window, the timeout operation
+ *    must be completed.
+ * 2. The final thread in the waiter list had insufficient data copied to/from
+ *    it to wake it, and its timeout had expired--ensure it wakes.
+ *
+ * @return 1 to stop further walking; 0 to continue walking
```
Comment (ceolin) [inline kernel/pipes.c:245]: Is there any case where it is needed ? All calls to this function ignores its return.

**Review by andyross — COMMENTED** [review]
Sorry, took a while to collect my thought here.

I don't hate this and won't give it a -1, but I don't know that I love it either.  Adding state to the thread struct is still clumsy.

If the core race is that a timeout can expire while you walk the list of threads to wake up, can't you just put a separate lock around the timeout processing and hold that in the pipes code?  Seems like that would be a cleaner API to support (e.g. z_sched_un/lock_timeouts(), which is acquired and released around each individual thread in the timeout) with more easily understood semantics.  Not sure if that's enough to treat the race, but it sounds like it to me.

**Comment by peter-mitsis** [comment]
@andyross - You may be right! That would be a far simpler solution. (Sometimes I get in a rut of patterned thinking that can be hard to get out of.) I'm going to explore this a bit just to make sure.

**Comment by peter-mitsis** [comment]
If I am following your idea correctly, somewhere in the z_thread_timeout() we would try to obtain the pipe spin lock. Since there is (currently) no way from z_thread_timeout() to determine the relevant pipe lock, this would likely be a global lock that we would blindly try to obtain each time. This would then suggest that pipe locks would no longer need to be on a per-pipe basis, but instead use this global pipe lock. Assuming that what I have described is something similar to what you envisioned, then the biggest drawback is that every timeout becomes linked with pipe operations. That linkage does not sit well with me.

I think that this drawback can be ameliorated. We currently keep a pointer to the wait queue in k_thread. If we also kept a pointer to a relevant spinlock in k_thread as well, then z_thread_timeout() could just lock on that (this would permit us to keep a per-pipe lock). For cases where we don’t really need a spin lock, it could point to a dummy lock (that always exists).

File: kernel/thread.c [hunk 2]
```diff
@@ -1126,9 +1130,27 @@ int k_thread_runtime_stats_all_get(k_thread_runtime_stats_t *stats)
...
 	struct k_thread *thread = CONTAINER_OF(timeout,
 					       struct k_thread, base.timeout);
 
+	/*
+	 * Wait until the spin lock associated with the object upon which
+	 * the thread was pended is available. If there was no such object
+	 * the spin lock would be z_obj_lock and can be expected to be
+	 * available. This provides simple protection against timeout induced
+	 * wait queue changes for kernel objects that must operate on multiple
+	 * waiting items.
```
Comment (andyross) [inline kernel/thread.c:1144]: Not sure this explanation is completely right.  The lock parameter to pend_curr() is essentially a condition variable, though it doesn't use the term.  The pend call atomically (!) releases the lock and puts the thread to sleep, so there's no intermediate state visible.  This change just makes the timeout processing part of the same paradigm: the timeout is guaranteed to see one of (1) the lock held and the thread still-unpended or (2) the lock available and the thread asleep.  In the case of (1), taking the lock then guarantees that we wait until the thread is pended.

File: kernel/thread.c [hunk 3]
```diff
@@ -39,6 +39,8 @@ LOG_MODULE_DECLARE(os, CONFIG_KERNEL_LOG_LEVEL);
 static struct k_spinlock z_thread_monitor_lock;
 #endif /* CONFIG_THREAD_MONITOR */
 
+static struct k_spinlock z_obj_lock;
```
Comment (andyross) [inline kernel/thread.c:42]: This seems kinda needless to me.  Surely the code will end up smaller if you just check the field for NULL, no?

File: kernel/sched.c [hunk 4]
```diff
@@ -827,6 +827,7 @@ int z_pend_curr(struct k_spinlock *lock, k_spinlock_key_t key,
 	 * held.
 	 */
 	(void) k_spin_lock(&sched_spinlock);
+	_current->obj_lock = lock;
```
Comment (andyross) [inline kernel/sched.c:828]: Is this not still racy?  You hold the scheduler lock here, but the unpend/timeout code that is going to end up reading this value does not.

**Review by andyross — COMMENTED** [review]
This technique seems to have promise, essentially promoting the swap/pend condition variable into an API for the timeout subsystem.  Some nitpicks on how to desribe it, and also a worry about a remaining race.

**Comment by nashif** [comment]
@peter-mitsis  what is thr status of this? Can you rebase please.

**Comment by peter-mitsis** [comment]
This has gone on the back burner while other items have taken higher priority.  I think that a better solution than this will require significant enough re-work that it would be worthwhile to do a new PR. Closing this one.

**Pull request closed without merging**